addons/website_sale_loyalty/controllers/main.py

```diff
@@ -0,0 +1,67 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import http, _
+from odoo.addons.website_sale.controllers import main
+from odoo.http import request
+
+class WebsiteSale(main.WebsiteSale):
+    def _prepare_cart_values(self, request, values, **post):
+        """ Publishes the available rewards to the cart view rendering """
+        super()._prepare_cart_values(request, values, **post)
+        if not post or post.get('type') != 'popover':
+            order = request.website.sale_get_order()
+            if order.partner_id and request.website.has_loyalty and request.website.loyalty_id \
+                    and not request.website.is_public_user():
+                values['available_rewards'] = request.website.loyalty_id.get_available_rewards(order)
+
+    def _get_shop_payment_values(self, order, **kwargs):
+        values = super()._get_shop_payment_values(order, **kwargs)
+        order = request.website.sale_get_order()
+        if order.available_loyalty_points < 0:
+            values['errors'].append((
+                _('Sorry, you do not have sufficient loyalty points for all rewards in your cart.'),
+                _('Please remove some of the rewards before proceeding with check out.'),
+            ))
+        return values
+
+    def checkout_form_validate(self, mode, all_form_values, data):
+        order = request.website.sale_get_order()
+        if order.available_loyalty_points < 0:
+            error = dict()
+            error['order_line'] = 'error'
+            error_message = [(_('Insufficient loyalty points: please remove some rewards.'))]
+            return error, error_message
+        return super().checkout_form_validate(mode, all_form_values, data)
+
+    @http.route()
+    def cart_update_json(self, product_id, line_id=None, add_qty=None, set_qty=None, display=True, **kw):
```

Comment:
Not quite, therefore a comment is needed indeed: by updating the cart, the number of consumed points might evolve, thus the available rewards might evolve. Since the client rpc call is only about updating the cart, if we notice that the update makes the list of rewards different, it returns a response that triggers a full page refresh so that the updated rewards list is displayed.